### PR TITLE
restrict increase, decrease, and burn and v4 initialize pool call

### PR DESCRIPTION
**Pull request by dianakocsis**

### CODE REVIEW

File: contracts/base/Dispatcher.sol [hunk 1]
```diff
@@ -269,9 +271,40 @@ abstract contract Dispatcher is Payments, V2SwapRouter, V3SwapRouter, V4SwapRout
...
                     (success, output) = address(V3_POSITION_MANAGER).call(inputs);
+                } else if (command == Commands.V4_INITIALIZE_POOL) {
+                    bytes4 selector;
+                    assembly {
+                        selector := calldataload(inputs.offset)
+                    }
+                    if (selector != PoolInitializer.initializePool.selector) {
+                        revert InvalidAction(selector);
+                    }
+                    (success, output) = address(V4_POSITION_MANAGER).call(inputs);
```
Comment (hensha256) [inline contracts/base/Dispatcher.sol:282]: What is the need for the call to go to the position manager? Wouldnt it be better to just call the pool manager directly? and instead of checking the selector just call `PoolManager.initialize(params)`? Or is there a reason to do so via the position manager?

File: contracts/base/Dispatcher.sol [hunk 2]
```diff
@@ -269,9 +271,40 @@ abstract contract Dispatcher is Payments, V2SwapRouter, V3SwapRouter, V4SwapRout
...
+                    bytes calldata slice = inputs[4:];
+                    (bytes calldata actions,) = slice.toBytes(0).decodeActionsRouterParams();
+
+                    uint256 numActions = actions.length;
+
+                    for (uint256 actionIndex = 0; actionIndex < numActions; actionIndex++) {
+                        uint256 action = uint8(actions[actionIndex]);
+
+                        if (
+                            action == Actions.INCREASE_LIQUIDITY || action == Actions.DECREASE_LIQUIDITY
```
Comment (hensha256) [inline contracts/base/Dispatcher.sol:302]: i think this should all be in a helper function thats called like `checkV4PositionManagerCall` inside `V3ToV4Migrator`, and then this contract can stay way simpler. And all this logic can just go into there. So here its just

```
} else if (command == Commands.V4_POSITION_CALL) {
    checkV4PositionManagerCall()
    (success, output) = address(V4_POSITION_MANAGER).call{value: address(this).balance}(inputs);
}
```

and i'd argue we shouldve done the same for v3 but that can be separate PR to refactor

File: contracts/base/Dispatcher.sol [hunk 3]
```diff
@@ -269,9 +271,40 @@ abstract contract Dispatcher is Payments, V2SwapRouter, V3SwapRouter, V4SwapRout
...
+                    bytes4 selector;
+                    assembly {
+                        selector := calldataload(inputs.offset)
+                    }
+                    if (selector != V4_POSITION_MANAGER.modifyLiquidities.selector) {
+                        revert InvalidAction(selector);
+                    }
+
+                    bytes calldata slice = inputs[4:];
+                    (bytes calldata actions,) = slice.toBytes(0).decodeActionsRouterParams();
```
Comment (hensha256) [inline contracts/base/Dispatcher.sol:294]: oh you never use the parameters its decoding... i feel like it would be possible instead (and more efficient) to try this...
`slice.toBytes(0).toBytes(0)`

File: contracts/modules/V3ToV4Migrator.sol [hunk 4]
```diff
@@ -3,10 +3,18 @@ pragma solidity ^0.8.24;
 
 import {MigratorImmutables} from '../modules/MigratorImmutables.sol';
 import {INonfungiblePositionManager} from '@uniswap/v3-periphery/contracts/interfaces/INonfungiblePositionManager.sol';
+import {Actions} from '@uniswap/v4-periphery/src/libraries/Actions.sol';
+import {CalldataDecoder} from '@uniswap/v4-periphery/src/libraries/CalldataDecoder.sol';
 
 /// @title V3 to V4 Migrator
 /// @notice A contract that migrates liquidity from Uniswap V3 to V4
 abstract contract V3ToV4Migrator is MigratorImmutables {
+    using CalldataDecoder for bytes;
```
Comment (dianakocsis) [inline contracts/modules/V3ToV4Migrator.sol:12]: this isn't necessary anymore right? since youre just doing .toBytes(0) instead of decodeActionsRouterParams()

File: contracts/base/Dispatcher.sol [hunk 5]
```diff
@@ -24,8 +26,6 @@ abstract contract Dispatcher is Payments, V2SwapRouter, V3SwapRouter, V4SwapRout
 
     error InvalidCommandType(uint256 commandType);
     error BalanceTooLow();
-    error InvalidAction(bytes4 action);
-    error NotAuthorizedForToken(uint256 tokenId);
```
Comment (hensha256) [inline contracts/base/Dispatcher.sol:28]: just moved to the v3 migrator module

File: contracts/modules/V3ToV4Migrator.sol [hunk 6]
```diff
@@ -20,4 +28,33 @@ abstract contract V3ToV4Migrator is MigratorImmutables {
...
+            selector := calldataload(inputs.offset)
+        }
+        if (selector != V4_POSITION_MANAGER.modifyLiquidities.selector) {
+            revert InvalidAction(selector);
+        }
+
+        bytes calldata slice = inputs[4:];
+        // first bytes(0) extracts the unlockData parameter from modifyLiquidities
+        // second bytes(0) extracts the actions parameter from unlockData
+        bytes calldata actions = slice.toBytes(0).toBytes(0);
```
Comment (snreynolds) [inline contracts/modules/V3ToV4Migrator.sol:45]: 🥵 tricky

**Review by snreynolds — COMMENTED** [review]
I think seems good to me, but I noticed in the tests that we dont actually use initialize before the mint, we just have a gas test for it. Unless I'm missing a test file? Could we add a full e2e test encoding the initialize then the mint, then check the success?

File: contracts/modules/V3ToV4Migrator.sol (hunk 6 above)
Comment (hensha256) [reply contracts/modules/V3ToV4Migrator.sol:50]: added more comments

**Review by snreynolds — APPROVED** [review]

File: contracts/modules/V3ToV4Migrator.sol [hunk 7]
```diff
@@ -20,4 +28,38 @@ abstract contract V3ToV4Migrator is MigratorImmutables {
...
+        // the second bytes(0) extracts the actions parameter from unlockData
+        bytes calldata actions = slice.toBytes(0).toBytes(0);
+
+        uint256 numActions = actions.length;
+
+        for (uint256 actionIndex = 0; actionIndex < numActions; actionIndex++) {
+            uint256 action = uint8(actions[actionIndex]);
+
+            if (
+                action == Actions.INCREASE_LIQUIDITY || action == Actions.DECREASE_LIQUIDITY
```
Comment (kmbarry1) [inline contracts/modules/V3ToV4Migrator.sol:58]: IMO it would be clearer (and should save some bytecode size) to express this condition as:

`if (action != Actions.MINT_POSITION) {`

Comment (kmbarry1) [reply contracts/modules/V3ToV4Migrator.sol:58]: Or is there an explicit desire to allow all the other actions, and just disallow these three? In that case I'd argue the error name is a bit misleading and be `ForbiddenAction()` or something like that.

Comment (kmbarry1) [reply contracts/modules/V3ToV4Migrator.sol:58]: Thinking about it more, it's gotta be the second case, so ignore the original comment and consider an error naming tweak.

File: contracts/base/Dispatcher.sol [hunk 8]
```diff
@@ -269,12 +269,21 @@ abstract contract Dispatcher is Payments, V2SwapRouter, V3SwapRouter, V4SwapRout
                     }
 
                     (success, output) = address(V3_POSITION_MANAGER).call(inputs);
-                } else if (command == Commands.V4_POSITION_CALL) {
+                } else if (command == Commands.V4_INITIALIZE_POOL) {
+                    PoolKey calldata poolKey;
+                    uint160 sqrtPriceX96;
+                    assembly {
+                        poolKey := inputs.offset
+                        sqrtPriceX96 := calldataload(add(inputs.offset, 0xa0))
```
Comment (kmbarry1) [inline contracts/base/Dispatcher.sol:277]: It may be nice to leave a comment explaining the `0xa0` constant, it's a somewhat fragile pattern.

File: contracts/modules/V3ToV4Migrator.sol (hunk 7 above)
Comment (hensha256) [reply contracts/modules/V3ToV4Migrator.sol:58]: Yeah because we still want settling, taking, clearing as actions. We can definitely make the error better and leave a comment to make this clearer

Comment (HickupHH3) [reply contracts/modules/V3ToV4Migrator.sol:58]: alternative check is `if (action < Actions.SETTLE && action != Actions.MINT_POSITION)`, not sure if it'll be cheaper
